api/api.proto

```diff
@@ -3229,6 +3238,73 @@ message SdkNodeInspectRequest {
   string node_id = 1;
 }
 
+// Options for draining volume attachment from a node
+message SdkNodeDrainAttachmentOptions {
+    // Selector is used for selecting volumes whose
+    // attachment needs to be moved from this node.
+    // The selector could be a list of volume label
+    // key value pairs to select a subset of volumes.
+    map<string, string> selector = 1;
+    // Drain only sharedv4 volumes from a node
+    // By default all volumes will be drained.
+    bool only_sharedv4 = 2;
+}
+
+// Options for draining volume replicas from a node
+message SdkNodeDrainVolumeReplicaOptions {
+    // Selector is used for selecting volumes whose
+    // replicas need to be moved from this node to another.
+    // The selector could be a list of volume label
+    // key value pairs to select a subset of volumes.
+    map<string, string> selector = 1;
+}
+
+// Options for draining/detaching cloud drives from a node
+message SdkNodeDrainCloudDriveOptions {
+    // When set the cloud drives will also get deleted.
+    bool delete_drives = 1;
+}
+
+// Options for draining kvdb from a node
+message SdkNodeDrainKvdbOptions {
+}
+
+
+// Options for draining a node in the cluster
+message SdkNodeDrainOptions {
+    // Perform all the node drain related operations.
+    // This is a single switch to perform all the supported
+    // node drain sub operations.
+    bool all = 1;
+    // Selector is a global selector used while performing
+    // a node drain sub operation. Per option selector take
+    // precedence over global selector
+    map<string, string> selector = 2;
+    // When set volume attachments will be drained from this node.
+    SdkNodeDrainAttachmentOptions drain_attachments = 3;
+    // When set volume replicas from this node will be moved to
+    // another node
+    SdkNodeDrainVolumeReplicaOptions drain_volume_replicas =  4;
```

Comment:
Not sure if we want to do this be default.

If we consider the PKS case, during drain, we will detach the cloud drives itself. So doing repl moves by default might be an overkill since it's upgrades one node at a time.

Reply:
A node drain operation could also involve just draining the replicas of a volume or just a subset of volumes from a node. Thats the reason why if an option is not set the Storage implementation will not perform that specific node drain operation. That also allows performing only certain sub drain operations.